File: typescript-sdk/src/patch-extractor.ts
import fs from 'fs';
import path from 'path';
import sharp from 'sharp';
import { BBox } from './types';

function clampBBox(bbox: BBox, width: number, height: number, padding: number): BBox {
  let [left, top, right, bottom] = bbox;
  left = Math.max(0, left - padding);
  top = Math.max(0, top - padding);
  right = Math.min(width - 1, right + padding);
  bottom = Math.min(height - 1, bottom + padding);
  if (right < left) right = left;
  if (bottom < top) bottom = top;
  return [left, top, right, bottom];
}

export async function extractPatch(
  imagePath: string,
  bbox: BBox,
  outputPath?: string,
  padding: number = 0
): Promise<{ source_image_path: string; bbox: BBox; patch_size: [number, number]; output_path: string | null }> {
  const image = sharp(imagePath).ensureAlpha();
  const metadata = await image.metadata();
  if (!metadata.width || !metadata.height) {
    throw new Error('invalid_image_dimensions');
  }
  const [left, top, right, bottom] = clampBBox(bbox, metadata.width, metadata.height, padding);
  const extractRegion = { left, top, width: right - left + 1, height: bottom - top + 1 };
  const patch = image.extract(extractRegion);

  let normalizedOutput: string | null = null;
  if (outputPath) {
    const target = path.resolve(outputPath);
    await fs.promises.mkdir(path.dirname(target), { recursive: true });
    await patch.png().toFile(target);
    normalizedOutput = target;
  }

  return {
    source_image_path: imagePath,
    bbox: [left, top, right, bottom],
    patch_size: [extractRegion.width, extractRegion.height],
    output_path: normalizedOutput,
  };
}

File: python-sdk/captcha_background_sdk/patch_extractor.py
from __future__ import annotations

from pathlib import Path
from typing import Optional, Tuple

from PIL import Image

BBox = Tuple[int, int, int, int]


def _clamp_bbox(bbox: BBox, width: int, height: int, padding: int = 0) -> BBox:
    left, top, right, bottom = bbox
    left = max(0, left - padding)
    top = max(0, top - padding)
    right = min(width - 1, right + padding)
    bottom = min(height - 1, bottom + padding)
    if right < left:
        right = left
    if bottom < top:
        bottom = top
    return (left, top, right, bottom)


def extract_patch(
    image_path: str,
    bbox: BBox,
    output_path: Optional[str] = None,
    padding: int = 0,
) -> dict:
    with Image.open(image_path) as image:
        rgba = image.convert("RGBA")
        width, height = rgba.size
        left, top, right, bottom = _clamp_bbox(bbox, width=width, height=height, padding=padding)
        patch = rgba.crop((left, top, right + 1, bottom + 1))

    normalized_output_path = None
    if output_path:
        target = Path(output_path)
        if target.parent and not target.parent.exists():
            target.parent.mkdir(parents=True, exist_ok=True)
        patch.save(target)
        normalized_output_path = str(target)

    patch_width, patch_height = patch.size
    return {
        "source_image_path": image_path,
        "bbox": (left, top, right, bottom),
        "patch_size": (patch_width, patch_height),
        "output_path": normalized_output_path,
    }
